extensions/objects/templates/math-editor.component.ts

```diff
@@ -25,38 +25,41 @@ angular.module('oppia').component('mathEditor', {
     value: '='
   },
   template: require('./math-editor.component.html'),
-  controller: [function() {
+  controller: ['$scope', function($scope) {
     const ctrl = this;
+    const SYMBOLS_TO_REMOVE = [
+      'norm', 'utf8', 'text', 'sym_name', 'eval', 'floor', 'factorial', 'sub',
+      'int', 'defi', 'deriv', 'sum', 'prod', 'root', 'vec', 'point',
+      'infinity', 'leq', 'less', 'geq', 'greater', 'neq'];
 
-    ctrl.assignIdToGuppyDiv = function() {
-      // Dynamically assigns a unique id to the guppy-div
-      var divId = 'guppy_' + Math.floor(Math.random() * 100000000);
-      $('.guppy-div').attr('id', divId);
-      return divId;
-    };
-
-    ctrl.createGuppyInstance = function(divId: string) {
-      return new Guppy(divId, {});
+    ctrl.initializeGuppy = function() {
+      var guppyDivs = document.querySelectorAll('.guppy-div');
+      var divId, guppyInstance;
+      for (var i = 0; i < guppyDivs.length; i++) {
+        divId = 'guppy_' + Math.floor(Math.random() * 100000000);
+        // Dynamically assigns a unique id to the guppy div.
+        guppyDivs[i].setAttribute('id', divId);
+        // Create a new guppy instance for that div.
+        guppyInstance = new Guppy(divId, {});
+        guppyInstance.event('change', (e) => {
+          ctrl.value = guppyInstance.asciimath();
+          // Need to manually trigger the digest cycle
+          // to make any 'watchers' aware of changes in ctrl.value.
+          $scope.$apply();
+        });
+        guppyInstance.configure('buttons', ['controls']);
+        guppyInstance.configure(
+          'empty_content',
+          '\\color{grey}{\\text{\\small{Type a formula here.}}}');
+      }
     };
 
     ctrl.$onInit = function() {
       ctrl.alwaysEditable = true;
-      var divId = ctrl.assignIdToGuppyDiv();
-      var guppyInstance = ctrl.createGuppyInstance(divId);
-      guppyInstance.event('change', (e) => {
-        ctrl.value = guppyInstance.asciimath();
-      });
-      guppyInstance.configure('buttons', ['controls']);
-      guppyInstance.configure(
-        'empty_content',
-        '\\color{grey}{\\text{\\small{Type a formula here.}}}');
+      ctrl.initializeGuppy();
 
       // Remove symbols since they are not supported.
-      var symbolsToRemove = [
-        'norm', 'utf8', 'text', 'sym_name', 'eval', 'floor', 'factorial', 'sub',
-        'int', 'defi', 'deriv', 'sum', 'prod', 'root', 'vec', 'point',
-        'infinity', 'leq', 'less', 'geq', 'greater', 'neq'];
-      for (var symbol of symbolsToRemove) {
+      for (var symbol of SYMBOLS_TO_REMOVE) {
         Guppy.remove_global_symbol(symbol);
```

Comment:
I'm a bit confused -- I thought that we had multiple components, but they all correspond to a single global guppy third-party lib.

So, if you do this, wouldn't Guppy.remove_global_symbol() be called multiple times if there are multiple components on the page? This is the sort of thing I'm worried about -- things that should be done once should only be done once and not multiple times. (Let me know if I'm missing something.)

Reply:
Oh, I see. I'm sorry I misinterpreted your question. So, to ensure that the removal code is only called once, we'd have to put it in a place that is only invoked once. Options for doing this are:
1. In guppy.html with a script tag.
2. A separate guppy_config.js that the guppy.html would include with the script src tag just like it includes the third party lib guppy.js.
3. A global file like App.ts. (I tried doing this but it didn't work. It gave an error saying that `Guppy is not defined.`. I think this is because of the order of loading of files -- the guppy.html file needs to be loaded before App.ts which is probably not happening.

Also, creating a separate service wouldn't help since we would have to inject that service in the component and the functions in there would have to be called in the onInit method which would effectively cause the same problem that is currently present, i.e., removal code would be called multiple times.

Reply:
Well, that's why I was saying that the separate service would have a boolean flag that checks whether init() has already been called or not. The first call would set it, and other calls would check whether the boolean has been set and, if so, return without re-executing the logic. Does that make sense / would that work?

(Also, in general, this doesn't just apply to the removal code. It applies to anything that should  be called only once in a page's lifetime.)

Reply:
That makes sense, yes. One question though -- can't we do that without using a separate service? I can just have a variable that checks if onInit has been called or not and based on that I can execute the removal and the initialization code. Does that seem reasonable?

Reply:
If my understanding of components is correct, that won't work. Each component is completely separate from the others and it's not very useful to check that its own onInit has been called or not. The point of the service is that it's a singleton across the entire app.

Also, make sure the service is fully tested!

Reply:
Yup, I tried that and it didn't work. I've added a service now, PTAL, thanks!